sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+                    if let Ok(interface_name) = device.read_string_descriptor_ascii(idx) {
+                        let ports = device.port_numbers()?.iter().map(
+                            |id| id.to_string()).collect::<Vec<String>>().join(".");
+                        let interface_path = path
+                            .join(format!("{}-{}", device.bus_number(), ports))
+                            .join(format!("{}-{}:{}.{}",device.bus_number(), ports,
+                                          config_desc.number(), interface.number()));
+                        // Check the ASCII name of this USB interface.
+                        if interface_name == "HyperDebug Shell" {
+                            // We found the "main" control interface of HyperDebug, allowing
+                            // textual commands to be sent, to e.g. manipoulate GPIOs.
+                            console_tty = Some(Self::find_tty(&interface_path)?)
+                        } else if interface_name.starts_with("UART") {
+                            // We found an UART forwarding USB interface.
+                            uart_ttys.insert(
+                                interface_name.clone(),
+                                Self::find_tty(&interface_path)?);
+                        } else if interface_name == "SPI" {
+                            // We found the SPI forwarding USB interface (this one interface
+                            // allows multiplexing physical SPI ports.)
+                            let mut in_endpoint: u8 = 0;
+                            let mut out_endpoint: u8 = 0;
+                            for endpoint_desc in interface_desc.endpoint_descriptors() {
+                                if endpoint_desc.transfer_type() == rusb::TransferType::Bulk {
+                                    match endpoint_desc.direction() {
+                                        rusb::Direction::In => {
+                                            in_endpoint = endpoint_desc.address();
+                                        }
+                                        rusb::Direction::Out => {
+                                            out_endpoint = endpoint_desc.address();
+                                        }
+                                    }
+                                }
+                            }
+                            if in_endpoint == 0 || out_endpoint == 0 {
+                                return Err(Error::CommunicationError("Missing SPI endpoint").into());
+                            }
+                            spi_interface.replace(BulkInterface {
+                                interface: interface.number(),
+                                in_endpoint,
+                                out_endpoint,
+                            });
+                        }
+                    }
+                }
+            }
+        }
+        let mut spi_names = HashMap::new();
```

Comment:
@cfrantz don't we have a nice macro for creating collections?

Reply:
Yes.  In opentitanlib, we have util:collection:

You should be able to write something like:
```
use crate::util::collection;

let spi_names = collection! {
  "SPI2" => 0,
  "0" => 0,
};
```

You might need to write `let spi_names<String, u32>: = collection!{ ... }` or add `.to_string()` to the strings to force them to be the correct type.

Reply:
Done